src/core/thread/mesh_forwarder.cpp

```diff
@@ -1345,6 +1342,13 @@ void MeshForwarder::HandleFragment(uint8_t *aFrame, uint8_t aFrameLength,
 
     // copy Fragment
     message->Write(message->GetOffset(), aFrameLength, aFrame);
+
+    if (fragmentHeader->GetDatagramOffset() == 0)
+    {
+        // Security Check (only if this is the first fragment)
+        VerifyOrExit(mNetif.GetIp6Filter().Accept(*message), error = kThreadError_Drop);
```

Comment:
The current `Message::Free()` will not dequeue the message for you.  As a result, this will result in freeing a message that is enqueued.  I think it's safer to keep this in the if-statement block above and just move the `message->Write()` call up into the if-statement as well.

Separately, I think it would be a good idea for `Message::Free()` to automatically dequeue the message.  I can look into that.